File: macos/Sources/FloatingLyricCore/Auth/SpotifyCallback.swift
import Foundation

/// Where Spotify sends the browser back to, and how to read what it sent.
///
/// The app registers a private URL scheme rather than a loopback address:
/// `ASWebAuthenticationSession` intercepts a custom scheme itself, which is
/// what lets the login happen in a sheet with no local web server, no port to
/// fight over, and no firewall prompt.
public enum SpotifyCallback {
    public static let scheme = "floatinglyric"
    public static let uri = "floatinglyric://callback"

    /// Reads the authorization code out of the URL Spotify redirected to,
    /// rejecting anything that does not carry back the `state` we sent.
    ///
    /// The state check is the CSRF defence: without it, someone else's
    /// authorization response would be accepted as if it were ours.
    public static func code(from url: URL, expectedState: String) -> Result<String, AppError> {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return .failure(.authCancelled)
        }
        let items = components.queryItems ?? []
        func value(_ name: String) -> String? {
            items.first { $0.name == name }?.value?.nilIfEmpty
        }

        // Spotify reports a refusal — "access_denied" when the user clicks
        // Cancel — in the query rather than as a failed request.
        if value("error") != nil { return .failure(.authCancelled) }

        guard let state = value("state"), state == expectedState else {
            return .failure(.authStateMismatch)
        }
        guard let code = value("code") else { return .failure(.authCancelled) }
        return .success(code)
    }
}

File: windows/src/FloatingLyric.Core/SpotifyCallback.cs
using System.Web;

namespace FloatingLyric.Core;

/// <summary>
/// Where Spotify sends the browser back to, and how to read what it sent.
///
/// Windows has no ASWebAuthenticationSession, so unlike the Mac build this
/// port uses the loopback redirect — which is why the Spotify dashboard needs
/// http://127.0.0.1:8888/callback registered.
/// </summary>
public static class SpotifyCallback
{
    public static readonly int[] CandidatePorts = [8888, 8889, 8890];

    public static string RedirectUri(int port) => $"http://127.0.0.1:{port}/callback";

    /// <summary>
    /// Reads the authorization code out of the callback, rejecting anything
    /// that does not carry back the state we sent — that is the CSRF defence,
    /// not decoration.
    /// </summary>
    public static Result Code(Uri url, string expectedState)
    {
        var query = HttpUtility.ParseQueryString(url.Query);
        string? Value(string name) =>
            string.IsNullOrEmpty(query[name]) ? null : query[name];

        // Spotify reports a refusal in the query, not as a failed request, and
        // sends no code with it — so this is checked before the state.
        if (Value("error") is not null) return Result.Fail(new AppError.AuthCancelled());
        if (Value("state") is not { } state || state != expectedState)
            return Result.Fail(new AppError.AuthStateMismatch());
        if (Value("code") is not { } code) return Result.Fail(new AppError.AuthCancelled());
        return Result.Ok(code);
    }

    public sealed record Result(string? Value, AppError? Error)
    {
        public static Result Ok(string code) => new(code, null);
        public static Result Fail(AppError error) => new(null, error);
        public bool IsSuccess => Error is null;
    }
}
